modules/currency.js

```diff
@@ -0,0 +1,220 @@
+
+import bidfactory from 'src/bidfactory';
+import { STATUS } from 'src/constants';
+import { ajax } from 'src/ajax';
+import * as utils from 'src/utils';
+import bidmanager from 'src/bidmanager';
+
+$$PREBID_GLOBAL$$.currency = setConfig;
+
+const DEFAULT_CURRENCY_RATE_URL = 'http://currency.prebid.org/latest.json';
+const CURRENCY_RATE_PRECISION = 4;
+
+var bidResponseQueue = [];
+var conversionCache = {};
+var currencyRatesLoaded = false;
+var adServerCurrency = 'USD';
+
+// Used as reference to the original bidmanager.addBidResponse
+var originalBidResponse;
+
+export var currencySupportEnabled = false;
+export var currencyRates = {};
+var currencyOverrides = {};
+
+export function setConfig(config) {
+  var url = DEFAULT_CURRENCY_RATE_URL;
+  if (typeof config.adServerCurrency === 'string') {
+    utils.logInfo('enabling currency support', arguments);
+
+    adServerCurrency = config.adServerCurrency;
+    if (config.conversionRateFile) {
+      utils.logInfo('currency using override conversionRateFile:', config.conversionRateFile);
+      url = config.conversionRateFile;
+    }
+    initCurrency(url);
+  } else {
+    // currency support is disabled, setting defaults
+    utils.logInfo('disabling currency support');
+    resetCurrency();
+  }
+  if (typeof config.currencyOverrides === 'object') {
+    currencyOverrides = config.currencyOverrides;
+  }
+}
+
+function initCurrency(url) {
+  conversionCache = {};
+  currencySupportEnabled = true;
+
+  if (!originalBidResponse) {
+    utils.logInfo('Installing addBidResponse decorator for currency module', arguments);
+
+    originalBidResponse = bidmanager.addBidResponse;
+    bidmanager.addBidResponse = addBidResponseDecorator(bidmanager.addBidResponse);
```

Comment:
oh yeah... I agree, a standardized way to implement plugins is way outside of the scope. I'm not even sure it's possible... but if it is, we definitely need more use-cases.

My suggestion would be something like this (not sure if you need sync or async functions... so I described both):

```
bidAdjustmentsRegistry.js:

function newBidAdjustmentRegistry() {
  const adjustments = []
  return {
    // adjuster should be a function of the type (bid) => bid, or (bid, callback) => void for async
    register(adjuster) {
      adjustments.push(adjuster);
    },
    adjust(bid) { // May need to be adjust(bid, callback) if async
      // execute each function in the queue, and return the result or do the callback. Return the input if none exist
    }
  }
}

export const bidAdjustments = newBidAdjustmentRegistry();
```

In the currency module, replace `bidmanager.addBidResponse = addBidResponseDecorator(bidmanager.addBidResponse);` with an equivalent `bidAdjustments.register(impl)`. I _think_ this is doable, but correct me if I'm wrong.

In #1421, @jaiminpanchal27 can call `bidAdjustments.adjust(bid)` inside the `addBidResponse` function. Or `bidAdjustments.adjust(bid, addBidResponse)` if it's async.

Does that sound like it'd work?